### DISCUSSION TITLE
IntegrityError trying to translate snippets

### DISCUSSION THREAD

**Discussion by enzedonline** [wagtail/wagtail-localize#483] [Q&A]
Not sure if this is a question or a bug ...

I've added a second locale to a site I'm working on, everything translates fine except one snippet class. I get an IntegrityError whenever I try to translate one of these snippets, regardless of whether it was existing pre-locale addition or new since.

It's a clusterable model with orderables, simple class, nothing special going on, it's just a method to drop translated static text into non-Wagtail pages on the site.

The only thing I can think of is that the clusterable model has no translatable fields (ie translatable_fields = []), the translatable fields are in the child orderable class. Would this trip this error?

I have another clusertable/orderable snippet class that translates fine, so this is the only real difference I see between the two.

The error I receive on a brand new instance with no orderable yet saved (makes no difference to the error, just eliminating possible causes):

```
IntegrityError at /admin/localize/submit/snippet/site_settings/templatetext/21/

duplicate key value violates unique constraint "site_settings_templatetext_template_set_0baf3836_uniq"
DETAIL:  Key (template_set)=(translatetest) already exists.

Request Method: POST
http://localhost/admin/localize/submit/snippet/site_settings/templatetext/21/?next=%2Fadmin%2Fsnippets%2Fsite_settings%2Ftemplatetext%2F
3.2.6
IntegrityError
duplicate key value violates unique constraint "site_settings_templatetext_template_set_0baf3836_uniq" DETAIL:  Key (template_set)=(translatetest) already exists.
```

The snippet models:

```python
class TemplateText(TranslatableMixin, ClusterableModel):
    template_set = models.CharField(
        unique=True,
        max_length=50,
        verbose_name="Set Name",
        help_text=_("The set needs to be loaded in template tags then text references as {{set.tag}}")
    )    

    translatable_fields = []
    
    panels = [
        FieldPanel("template_set"),
        MultiFieldPanel(
            [
                InlinePanel("templatetext_items"),
            ],
            heading=_("Text Items"),
        ),
    ]

    def __str__(self):
        return self.template_set
    
    class Meta:
        verbose_name = _('Template Text')
        unique_together = ('translation_key', 'locale')

class TemplateTextSetItem(TranslatableMixin, Orderable):
    set = ParentalKey(
        "TemplateText",
        related_name="templatetext_items",
        help_text=_("Template Set to which this item belongs."),
        verbose_name="Set Name",
    )
    template_tag = models.SlugField(
        max_length=50,
        help_text=_("Enter a tag without spaces, consisting of letters, numbers, underscores or hyphens."),
        verbose_name="Template Tag",
    )    
    text = models.TextField(
        null=True,
        blank=True,
        help_text=_("The text to be inserted in the template.")
    )

    translatable_fields = [
        TranslatableField('text'),
    ]

    panels = [
        RegexPanel('template_tag', '^[-\w]+$'),
        FieldPanel('text'),
    ]

    def __str__(self):
        return self.template_tag
```

Any help greatly appreciated!

Django 3.2.6
Wagtail 2.14.1
Wagtail-localize 0.9.5

**Comment by zerolab** [comment]
I suspect what is happening here is your `template_set`'s unique constraint gets in the way of creating a new snippet version in the new locale. When you start a translation, a copy is created which has the same values as the original.

One way around it is to drop the unique there and set a `unique_together` constraint in the snippet meta class, 

`TranslatableMixin` uses `unique_together = [("translation_key", "locale")]` ([ref](https://github.com/wagtail/wagtail/blob/093d59d4e3f8057f4bdb281f9d8cf223c90cf0c5/wagtail/core/models/i18n.py#L93)) so you would probably want it as something like `[("translation_key", "locale", "template_set")]`

Note: this is all off the top of my head

**Reply by enzedonline** [reply]
Hi @zerolab , thanks for the reply.

I tried adding template_set to the unique_together property but it threw an error:
```
site_settings.TemplateText: (wagtailcore.E003) site_settings.TemplateText is missing a unique_together constraint for the translation key and locale fields
        HINT: Add ('translation_key', 'locale') to TemplateText.Meta.unique_together
```
Oddly, upgrading from 0.9.5 to 1.0.0 has removed the integrity error but now gives me other spurious behaviour. 

I get past the translate page ok:
![image](https://user-images.githubusercontent.com/72310199/140486210-5e242b6d-3829-4498-8095-8ac5d99eadfd.png)
But when I look on the Spanish page for the snippet, no objects have been created.
Back on the English page, the snippet instance still has the Translate option, but also the Sync Translated Snippets button.
![image](https://user-images.githubusercontent.com/72310199/140484277-11a51659-f841-4b6b-97e4-7ec82c853f52.png)

If I try to sync, it throws another error:
```
DoesNotExist at /admin/localize/update/23/
TemplateText matching query does not exist.
```

I tried removing the translatable_fields to see if that was tripping it, also adding a dummy translateable charfield. Neither of these have any effect.

I've also tried deleting the locale and starting over again.

Seems the only way around is to drop the unique as suggested and use 
`unique_together = ('translation_key', 'locale'), ('locale', 'template_set')`

However if a template_set is created with a duplicate name, the thrown error is unhandled by Wagtail:
```
duplicate key value violates unique constraint "site_settings_templatetext_locale_id_template_set_d0ec61a2_uniq"
DETAIL:  Key (locale_id, template_set)=(1, test5) already exists.
```

I can trap the error in a custom save method, though not sure how I return the user back to the form for amending, I guess there's a method for that? 

Trapping in the clean method requires knowing the locale of the snippet instance being edited which gets messy and a bit flaky since the only way I can find is to scrape it from the url (locale=xx in the query string for new instance, or look up the locale from the pk of the snippet in the last part of the path for existing instance).

Looks like wagtail-localize needs a way to handle unique constraints on snippets.

**Reply by zerolab** [reply]
Hey @enzedonline,

Looks like this issue is why we don't yet cant translate settings at the moment.
@kaedroho reckons the only way we'd support this is to translate the `unique` constraint to `unique_together` and maybe add a system check for that.

--
the other spurious behaviour from upgrading to 1.0.0 is interesting. Does it happen with other snippets or only this one?

**Reply by enzedonline** [reply]
@zerolab  Hmm, there's another item for my wish list: being able to access the instance locale of the object being created/edited - the self.usage_url() method doesn't return the querystring, so there's no way from that to get the locale_id for new instances. Since the request object isn't accessible during the clean method, maybe the only option currently is to use a custom edit handler that makes it a property of the field? This would be super useful for dynamic drop-downs that only show those options relevant to the instance locale, I'm scraping the URL for that presently.

This is the only one it happens on, but it's also the only one with the unique constraint. When I took that off, it behaved as normal. 

I have a test wagtail-localize with a couple of clusterable snippets, I'll add a unique constraint to one and see if it replicates.

**Reply by zerolab** [reply]
Can you please open an issue, if it doesn't exist?
Any PRs are welcome too ;)

**Reply by enzedonline** [reply]
Sure, will do.

I just spun up the test site and changed one of the fields to unique. It was on 0.9.5 and gave me the same constraint unhandled error. Upgraded to 1.0.0 and then got the same issue as described with the sync and translate buttons (with no translated copy made):

![image](https://user-images.githubusercontent.com/72310199/140529673-d5f5beb4-e2d7-4363-bf0c-0172b7b07d68.png)

Model for that one (with title field set to unique=true):

```
@register_snippet
class Menu(TranslatableMixin, ClusterableModel):
    """ Menu Class creates menus to display - validation in MenuForm
        Holds a collection of Menu Item orderables """
    
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)

    base_form_class = MenuForm

    title = models.CharField(
        max_length=50,
        unique=True,
        help_text=_("Title will be used if this is a submenu")
    )

    # Optional image to display if submenu
    icon = models.ForeignKey(
        "wagtailimages.Image",
        blank=True,
        null=True,
        on_delete=models.SET_NULL,
        related_name="+",
        help_text=_("Optional image to display if submenu")
    )
    
    panels = MenuPanelsIterable()

    def __str__(self):
        return self.title

```

**Reply by enzedonline** [reply]
Also occurring on standard Model snippets:

```
@register_snippet
class CompanyLogo(TranslatableMixin, models.Model):
    name = models.CharField(max_length=250, unique=True)
    logo = models.ForeignKey(
        "wagtailimages.Image", on_delete=models.CASCADE, related_name="+"
    )

    panels = [
        FieldPanel("name", classname="full"),
        ImageChooserPanel("logo"),
    ]

    def __str__(self):
        return self.name
```
